### DISCUSSION TITLE
Let's Improve!

### DISCUSSION THREAD

**Discussion by dmarro89** [dmarro89/dare-db#1] [Ideas]
Hi!
This project was born as an experimental project in Go but I want to improve it and make it complete.
Let's discusse some Ideas on how to improve the database.
Here's some points.
I'm using my [hashtable](https://github.com/dmarro89/go-redis-hashtable). Here a [discussion](https://github.com/dmarro89/go-redis-hashtable/issues/1) about how to improve its performances.
Then, what's next steps for a database?
- Any other data structure to use and commands to implement?
- Security - What's the best way to implement security on a DB ? Is the HTTP server the best choice ?
- Persistence - Understand how to make it real

You contribution is really appreciated.
Thanks

**Comment by vdmitriyev** [comment]
hi @dmarro89 

I am interested in your database project and would like to contribute, if you don't mind it. I also got some thought/ideas on security consideration for the database (connection, user/role management, secure data exchange, etc.). But I bet this particular discussion requires an additional thread.

**Reply by dmarro89** [reply]
Hi @vdmitriyev thanks for your comment. I'll create some threads about the security and I'll link them here.

**Reply by dmarro89** [reply]
https://github.com/dmarro89/dare-db/discussions/5
https://github.com/dmarro89/dare-db/discussions/4
https://github.com/dmarro89/dare-db/discussions/3

Here's some topics to start to think. Feel free to propose new discussions.
Thanks again.

**Reply by vdmitriyev** [reply]
thank you, I wrote some of my tough down in the discussions threads you have created

**Reply by dmarro89** [reply]
Ok, I've created some issues about how can be implemented.
Please feel free to contribute also with code.

**Reply by vdmitriyev** [reply]
thanks, I will do my best and going to try to contribute with the code as well

**Comment by vdmitriyev** [comment]
@dmarro89 I bet there is also another improvement requires for the database - configuration file of the database instance. Could I ask you to open another discussion thread for this?

**Reply by dmarro89** [reply]
Yes of course.

**Reply by vdmitriyev** [reply]
Link a discussion here - https://github.com/dmarro89/dare-db/discussions/8

**Reply by vdmitriyev** [reply]
I had created an issue to implement "configuration file" feature https://github.com/dmarro89/dare-db/issues/9 
However, I couldn't assign this issue to myself

**Comment by vdmitriyev** [comment]
I had created an initial draft of an architecture. It is in a ```mermaid``` format. Should we have a different discussion thread on that as well? Here is an architecture:

```mermaid
graph LR
    subgraph connection
        HTTPS(HTTPS)
        SOC(socket)
    end

    subgraph auth
        HTTPS --> RBAC(RBAC)
        SOC --> RBAC(RBAC)
    end
    
    subgraph engine
        ENG --> DAT[(data)]
        RBAC --> ENG(engine)
    end

    subgraph filesystem
        ENG --> CONF(config.yml) 
    end 
    
```

**Reply by dmarro89** [reply]
Yes of course, I'll create a new thread.
Thanks for the collaboration.

**Reply by dmarro89** [reply]
@vdmitriyev I've invited you as collaborator of the project

**Reply by vdmitriyev** [reply]
Thank you for trusting me with that. Really appreciate it very much

**Comment by vdmitriyev** [comment]
For completeness, I would like to link the idea about CLI here in the main discussion thread https://github.com/dmarro89/dare-db/issues/14

**Reply by vdmitriyev** [reply]
Discussion on CLI/ interactive shell with additional information on parameters and commands https://github.com/dmarro89/dare-db/discussions/16

**Comment by go-while** [comment]
based on this https://github.com/dmarro89/dare-db/issues/21

i fetched latest commits from develop branches into my playground branch and fixed the randomness stuff

can anyone explain why this is there? the idea was to feed some random bytes at boot time?

**Reply by dmarro89** [reply]
@go-while hi, I've read your comments. Thanks for your feedback.
The random bytes stuff is in another project -> https://github.com/dmarro89/go-redis-hashtable
It is used to generate a unique key for all the instances -> this key will be used during the siphash 2-4 digest generation.
This is the reason why a unique random bytes is generated at the start of the application.

**Reply by go-while** [reply]
Did I look in the wrong branch?
Can you point me where you initialize it?  
in my version the randomBytes were not initialized and always returned zero.

**Comment by vdmitriyev** [comment]
@dmarro89 should me write a short notice for contributors/developer (e.g., ```for the latest changes please check the develop branch```)?

**Comment by go-while** [comment]
anyways if it was initialized anywhere, it wasnt in my version. i debugged it hard and hashkey was mostly empty as well as randomBytes.
i changed a bunch of it and moved stuff from var into the Dict pointer, removed the needless split function evertyime. i've made it do it once and store key0, key1 in Dict too as they won't change on runtime.
but my merge fails, i somehow fu...d up with all the dependencies and GOPATH/src and modules f... around...

**Comment by go-while** [comment]
ok i believe i got it somehow, still not sure if everything imports correctly 

merged patches: go-redis-hashtable [(click)](https://github.com/go-while/go-redis-hashtable/commit/4030aee37b7520412dd77b619d0d631a2392c778)

merged patches: dare-db in branch [test-merge-02](https://github.com/go-while/dare-db/commits/test-merge-02/) 
merged all in one because deps killed my merge from playground branch... 
see [(commits history)](https://github.com/go-while/dare-db/commits/playground/)  here.

```bash
export LOGLEVEL=DEBUG
```
and run it, should see something new

**Reply by vdmitriyev** [reply]
I didn't get you. Are you speaking about [dare-db](https://github.com/dmarro89/dare-db/) or [go-redis-hashtable](https://github.com/dmarro89/go-redis-hashtable) ?

**Reply by dmarro89** [reply]
> ok i believe i got it somehow, still not sure if everything imports correctly 
> 
> 
> 
> merged patches: go-redis-hashtable [(click)](https://github.com/go-while/go-redis-hashtable/commit/4030aee37b7520412dd77b619d0d631a2392c778)
> 
> 
> 
> merged patches: dare-db in branch [test-merge-02](https://github.com/go-while/dare-db/commits/test-merge-02/) 
> 
> merged all in one because deps killed my merge from playground branch... 
> 
> see [(commits history)](https://github.com/go-while/dare-db/commits/playground/)  here.
> 
> 
> 
> ```bash
> 
> export LOGLEVEL=DEBUG
> 
> ```
> 
> and run it, should see something new
> 
> 
> 
> 
> 
> 

Hi @go-while, I'll prefer to talk about go-redis-hashtable in its github project.
In dare-db it is just used a dependency.
Thanks

**Reply by go-while** [reply]
sure but dare-db got changes which won't work without patching go-redis-hashtable

just created a pull in the other project

**Comment by go-while** [comment]
not yet ready to create a pull. just saw server.go is missing the logger functionality

**Comment by go-while** [comment]
https://github.com/go-while/dare-db/commit/226e52259429b65c250d683f97780582d181b9dd <- missed new logger

**Reply by vdmitriyev** [reply]
it was missing, because only a rudimental logging was added so far. A further logging supposed to be added in a different PR later

**Reply by vdmitriyev** [reply]
current implementation uses logrus, but future implementation should be using log/slog + context

**Comment by go-while** [comment]
X  Can’t automatically merge.  .. why it ?

**Reply by vdmitriyev** [reply]
Are you speaking about this PR https://github.com/dmarro89/dare-db/pull/22 ? 

It looks like you got a number of conflicts to resolve. Could you also provide further information on your PR? It is not quite clear, what is going on inside it.

**Comment by go-while** [comment]
see [(commits history)](https://github.com/go-while/dare-db/commits/playground/) here.

**Reply by go-while** [reply]
this PR was just a test to see what git does, because it showed an error on creating to pull, so i set it to draft

the go-redis-hashtable needs PR too or stuff won't work together.

i fixed the randomness, hashkey. was zero/empty when i debugged it. 
removed the split func. the randomBytes were not initialized as they should have been.
removed the logrus dep and implemented own logger as "darelog" :dancers:  with correct string formating
passed darelog to all instances, creates in main
export LOGLEVEL=DEBUG|INFO|WARN works as intended
removed mutexes from dare-db get/set/del and moved these to go-redis-hashtable, where they should be.

and for sure something more but i'm tired now :dango:

**Reply by go-while** [reply]
> current implementation uses logrus, but future implementation should be using log/slog + context

look deeper into the new logger i made ;)

**Reply by vdmitriyev** [reply]
thanks for the info. I put your summary into your pull request https://github.com/dmarro89/dare-db/pull/22

**Comment by go-while** [comment]
I got my develop branch to ~150k - 200k request / sec (insert, check) with my own tcp socket protocol and the client, not using http(s).
Needs some, ... or loads of ...[changes](https://github.com/go-while/nodare-db-dev) to the codebase to support more dicts at the first place.
I created an additional "idx" from the key to create  10, 100, 1000, 10000, 100000, 1000000, 10000000, 100000000, 100000000,  1000000000 or 4000000000 independent SubDICKS ^^ ...
... which can lock and rehash on it's own. more subdicks: more distributes the keys, less overloads the dicts.
 result: much faster baby 

+ imported go-redis-hashtable into the codebase: database/
+ Provides own logger.
+ Completely rewrote the boot sequence in main to have configs and logs early.
+ Added gorilla mux as http router.
+ raw TCP / TLS sockets with own protocol.
+ client test-app + clilib to talk to raw TCP / TLS sockets.

but because it's way to much. i won't bother you trying to understand all the changes in a pull request... 

but you can have a look and get ideas what you might need to boost performance ;) 

feel free to copy and paste. you're credited in the LIC!

**Reply by dmarro89** [reply]
> I got my develop branch to ~150k - 200k request / sec (insert, check) with my own tcp socket protocol and the client, not using http(s).
> 
> Needs some, ... or loads of ...[changes](https://github.com/go-while/nodare-db-dev) to the codebase to support more dicts at the first place.
> 
> I created an additional "idx" from the key to create  10, 100, 1000, 10000, 100000, 1000000, 10000000, 100000000, 100000000,  1000000000 or 4000000000 independent SubDICKS ^^ ...
> 
> ... which can lock and rehash on it's own. more subdicks: more distributes the keys, less overloads the dicts.
> 
>  result: much faster baby 
> 
> 
> 
> + imported go-redis-hashtable into the codebase: database/
> 
> + Provides own logger.
> 
> + Completely rewrote the boot sequence in main to have configs and logs early.
> 
> + Added gorilla mux as http router.
> 
> + raw TCP / TLS sockets with own protocol.
> 
> + client test-app + clilib to talk to raw TCP / TLS sockets.
> 
> 
> 
> but because it's way to much. i won't bother you trying to understand all the changes in a pull request... 
> 
> 
> 
> but you can have a look and get ideas what you might need to boost performance ;) 
> 
> 
> 
> feel free to copy and paste. you're credited in the LIC!
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 

Thanks @go-while

**Comment by go-while** [comment]
[socket.go](https://github.com/go-while/nodare-db-dev/blob/main/server/socket.go)

[client/clilib/client.go](https://github.com/go-while/nodare-db-dev/blob/main/client/clilib/client.go)

**Comment by go-while** [comment]
reading os signal is now single instance in main. with stop_chan below. stop_chan is passed to httpserver and waits there.
os.signal sends notify to stop_chan and every waiter just puts it back it so all waiters will eventually stop somehow. not great but works xD

**Comment by go-while** [comment]
the socket protocol is just a copy of how usenet transfers articles since decades.

I removed lots of headers from web-server too..  in "func [nilheader](https://github.com/go-while/nodare-db-dev/blob/3c9815622e700f4a8c6ebc003dfb89a8781ccf1d/server/ndb-server.go#L160C1-L160C14)"
Since a http client does not realy read them and does not read them with a Get only request, but server sends them.
I still get the "Transfer-Encoding: chunked" delivered for every http request which adds useless bytes to every request...

**Reply by dmarro89** [reply]
@go-while thanks for your work.
But I don't understand one thing.
Why you're not able to raise a PR over the develop branch with your changes?
Honestly, I appreciate your contribution but it's hard to analyze your change and compare them with the current code. It will be hard also to move them inside the current code. 

To make it easy, please raise a working PR able to pass all the checks in the action.
In this way me and @vdmitriyev are able to review and merge it.
Otherwise it'll be really hard to analyze it and honestly I don't have too much time to do it.

**Reply by go-while** [reply]
you won't be able to merge it as it is almost a complete rewrite.

**Reply by dmarro89** [reply]
Ok, I don't think that yours is the right approach.
You can contribute with pieces of code that can improve the performances but not rewrite a project.
Or at least contribute to the discussions (without spamming it with confusing messages).
Please, if you want to contribute I really appreciate it but I must ask you to align to the project rules.

**Reply by go-while** [reply]
i don't write go tests while working, so there are none. once there is a state reached that is worth to be testable by go, they might be added. or not. i like real world tests, like the client test app does, to test stuff in real time with real problems.

**Reply by go-while** [reply]
ok have fun my friend fixing your single thread app.

**Reply by dmarro89** [reply]
Also if is a complete rewrite, is possible to open a working PR to the develop branch. But I think that you don't know how to do it.
If you write code without testing it, please let me say that you're a junior profile needing some experience.
Really appreciate your contribution, if you want you can open a PR and your code will be analyzed.
Thanks

**Reply by dmarro89** [reply]
The app is a WIP and will change in time, step by step.
Enjoy with understanding how to open a PR.

**Reply by go-while** [reply]
> junior profile

ok my dictator like i said: have fun. i can work faster without you ;)

**Reply by dmarro89** [reply]
Thanks, bye

**Comment by dmarro89** [comment]
Too much waste here 👯
